pkg/operator/ceph/cluster/watcher_test.go

```diff
@@ -21,17 +21,106 @@ import (
 	"testing"
 
 	"github.com/coreos/pkg/capnslog"
+	"github.com/pkg/errors"
 	cephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+	rookv1 "github.com/rook/rook/pkg/apis/rook.io/v1"
 	"github.com/rook/rook/pkg/client/clientset/versioned/scheme"
 	"github.com/rook/rook/pkg/clusterd"
 	"github.com/rook/rook/pkg/operator/k8sutil"
+	exectest "github.com/rook/rook/pkg/util/exec/test"
 	"github.com/stretchr/testify/assert"
 	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client/fake"
 )
 
+func TestOnK8sNode(t *testing.T) {
+	ns := "rook-ceph"
+	// Register operator types with the runtime scheme.
+	s := scheme.Scheme
+	s.AddKnownTypes(cephv1.SchemeGroupVersion, &cephv1.CephCluster{})
+
+	cephCluster := &cephv1.CephCluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      ns,
+			Namespace: ns,
+		},
+		Status: cephv1.ClusterStatus{
+			Phase: "",
+		},
+		Spec: cephv1.ClusterSpec{
+			Storage: rookv1.StorageScopeSpec{
+				UseAllNodes:            false,
+				Nodes:                  nil,
+				VolumeSources:          nil,
+				StorageClassDeviceSets: nil,
+			},
+		},
+	}
+
+	object := []runtime.Object{
+		cephCluster,
+	}
+
+	// Create a fake client to mock API calls.
+	cl := fake.NewClientBuilder().WithScheme(s).WithRuntimeObjects(object...).Build()
+
+	executor := &exectest.MockExecutor{}
+	executor.MockExecuteCommandWithOutputFile = func(command, outputFile string, args ...string) (string, error) {
+		return "", errors.Errorf("failed to get osd list on host.")
+	}
+
+	clientCluster := newClientCluster(cl, ns, &clusterd.Context{
+		Executor: executor,
+	})
+
+	node := &corev1.Node{
+		Spec: corev1.NodeSpec{
+			Unschedulable: false,
+		},
+		Status: corev1.NodeStatus{
+			Conditions: []corev1.NodeCondition{
+				{
+					Status: corev1.ConditionStatus(corev1.ConditionTrue),
+					Type:   corev1.NodeConditionType(corev1.NodeReady),
+				},
+			},
+		},
+	}
+
+	// node not ready for reconcile
+	b := clientCluster.onK8sNode(node)
+	assert.False(t, b)
+
+	// node ready for reconcile
+	fakeNodes := []rookv1.Node{
+		{
+			Name: "nodeA",
+		},
+	}
+	fakeVolumes := []rookv1.VolumeSource{
+		{
+			Name: "volumeA",
+		},
+	}
+	fakeDeviceSets := []rookv1.StorageClassDeviceSet{
+		{
+			Name: "DeviceSet1",
+		},
+	}
+	nodesCheckedForReconcile.Add(node.Name) // adding node to nodesCheckedForReconcile Set
```

Comment:
Two concerns here:
1. IMO, its not a good practise to update global variables from the test method like this.  Let the actual method `onK8sNode` update `nodesCheckedForReconcile`.  
2. Also, the `node.Name` is empty because node name is not specified in the object above, so it doesn't look like this line of code working as expected.

Reply:
The `nodesCheckedForReconcile.Add(node.Name)` was added with an empty string.